### ISSUE TITLE
Slow/no database index(es), `logs` table

### ISSUE THREAD

**Issue by paalbra** [MISP/MISP#7945]
I'm experiencing *very* slow load times on some pages. E.g. `/logs/event_index/7380`.

After a look in the database I've noticed that data on this page comes from the `logs` table and that the event id is in the `model_id` column.

Why is there no index on this column?

A query like `select * from logs where model_id = 7380;` takes maybe 10-20 seconds.

I added an index to `model_id` (`create index model_id on logs (model_id);`) and the same query now takes 0.001 seconds.

Other pages concerning logs might still be a bit slow, but at least the event history is much faster.

My `/servers/serverSettings/diagnostics` page now complains since I have this extra index.

### Work environment

Containerized (podman) with MySQL 8.0 and MISP 2.4.150.

### Expected behavior

Decent load times.

### Actual behavior

Many seconds/up to a minute load or maybe timeouts.

### Steps to reproduce the behavior

Go to an event, click `View Event History`, og go direct via something like `/logs/event_index/7380`.

### Other info

I assume that this problem is present in my MISP instance due to a very large `logs` table.

```
MySQL [misp]> select count(*) as count, model from logs group by model order by count desc limit 5;
+---------+------------+
| count   | model      |
+---------+------------+
| 7834251 | User       |
|  470910 | Attribute  |
|  370118 | Server     |
|   43346 | MispObject |
|   25998 | Event      |
+---------+------------+
5 rows in set (13.940 sec)

MySQL [misp]> select count(*) from logs;
+----------+
| count(*) |
+----------+
|  8783285 |
+----------+
1 row in set (2.488 sec)
```

I'm thinking that I might have missed some kind of pruning/decay setting somewhere that removes old entries? I'm happy if anyone could guide me to documentation regarding this, if it exists.

Even though pruning this table could resolve my apparent issue I still do think one or multiple indexes would be nice to add to this table.

**Comment by paalbra** [comment]
I just found https://github.com/MISP/MISP/issues/6244

I have 7686808 user auths, AKA 88% of the table.

But I still do think it would be nice to add some indexes that could speed up things..

**Comment by iglocska** [comment]
Yeah absolutely. A lot of the tables lack some indexes simply for historical reasons (for example a field that originally wasn't a target for any searches) - the problem with adding indexes in update scripts after the fact is the utterly massive amount of time / space an update might end up taking if you have a long running / heavily used MISP instance. 

With that said, feel free to index unindexed fields to improve your performance, we'll be doing a larger change at one point that will require a manual update process and the plan is to tie some of these long running queries into that update.

**Comment by imidoriya** [comment]
@iglocska do you have some base recommendations for indexing that we could start using?  My performance is extremely slow and tips like this might be a great start.  Also, what can we do to reduce the size of the logs table.  Mine (after deleting a couple years) is still over 100Gig.  I'd turn it off completely if not for the warning that it's used in the events.  Seems like logging should be separate from required event data.

**Comment by paalbra** [comment]
> Seems like logging should be separate from required event data.

This seems like a good idea to me :+1:

**Comment by JoePJisc** [comment]
Adding my name to this too, one item we are seeing slow performance on is using the `/attributes/restSearch` endpoint to find attributes with specified tags (event or attribute tags), in a given time period and export as RPZ zones.

Any suggestions on additional indexes that would help this, would be really appreciated.

**Comment by imidoriya** [comment]
@JoeP-oss I think what I did was reviewed the log types and determined that my main issue was API Auth logs.  So I disabled them in settings as it wasn't overly important to me or the event history.

I also added an index to the date column in events and the model_id column in logs.

**Comment by dougjpsantos** [comment]
This is mine:

MariaDB [misp]>  select count(*) as count, model from logs group by model order                          by count desc limit 5;
+----------+-----------------+
| count    | model           |
+----------+-----------------+
| 49669282 | User            |
| 28578428 | Attribute       |
| 15111759 | MispObject      |
|  4464814 | Event           |
|    74245 | ObjectReference |
+----------+-----------------+
5 rows in set (20 min 44.094 sec)

MariaDB [misp]> select count(*) from logs;
+----------+
| count(*) |
+----------+
| 97986441 |
+----------+
1 row in set (22 min 29.065 sec)

MariaDB [misp]>

How do we add indexes to these tables ?

**Comment by JoePJisc** [comment]
These are the initial steps I've taken, based on the search criteria of our RPZ export, I'm also going to look to get slow query logging enabled, including the searches without indexes option, to detect any additional optimisations I can made.

```sql
CREATE INDEX to_ids ON attributes(to_ids);
CREATE INDEX published ON events(published);
CREATE INDEX timestamp ON events(timestamp);
```

With regards to logs, I've added this index and plan to schedule these deletion queries daily.

```sql
CREATE INDEX created ON audit_logs(created);

DELETE FROM audit_logs WHERE created < NOW() - INTERVAL 90 DAY;
DELETE FROM logs WHERE model='User' AND action='auth' AND created < NOW() - INTERVAL 90 DAY;
```

I'd welcome any thoughts on this / additional options

**Comment by iglocska** [comment]
I mean the logs are really only used by the event history - whether you care about that or not is a different question. I see some value in it when it comes to seeing the evolution of events, but not to the detriment of performance, so I think your strategy of purging older logs is sound.

**Comment by iglocska** [comment]
With that said - logs should never affect searches, so if you are having slow exports that is unrelated to this. It can be down to a bunch of different factors, currently we're in the process of finalising a release that should help with anything related to correlations for example. 

Make sure your DB configuration is up to snuff (especially innodb buffer size is crucial - without it you'll end up row scanning rather than using indeces). 

Other than that, optimising your export queries might help too, if you have some sample queries that are slow for you we can perhaps guide you in the right direction.
